Allow configuring unstable flags via config file

bearcage opened the pull request:
# Summary

This fixes #8127 by mapping the `unstable` key in `.cargo/config` to Z flags.

It should have no impact on stable/beta cargo, and on nightlies it gives folks the ability to configure Z flags for an entire project or workspace. This is meant to make it easier to try things like the [new features resolver](https://github.com/rust-lang/cargo/issues/8088) behavior, mtime-on-use, build-std, or timing reports across a whole project.

I've also included a small (but entirely independent) ergonomics change -- treating dashes and underscores identically in Z flags. That's along for the ride in this PR as the last commit, and is included here because it makes for more idiomatic toml file keys (`print_im_a_teapot = yes` vs `print-im-a-teapot = yes`). Please let me know if y'all would prefer that be in a separate PR, or not happen at all.

# Test Plan

Apologies if I've missed anything -- this is my first cargo contrib and I've tried to hew to the contributing guide. If I've slipped up, please let me know how and I'll fix it.

NB. My linux machine doesn't have multilib set up, so I disabled cross tests.

 * `cargo test` passes for each commit in the stack
 * I formatted each commit in the stack with `rustfmt` 
 * New tests are included alongside the relevant change for each change
 * I've validated each test by locally undoing the code change they support and confirming failure.
 * The CLI wins, for both enable and disabling Z flags, as you'd expect.

Keys in `unstable` which do not correspond with a Z flag will trigger an error indicating the invalid flag came from a config file read:

```
Invalid [unstable] entry in Cargo config

Caused by:
  unknown `-Z` flag specified: an-invalid-flag
```

If you'd like to see a test case which isn't represented here, I'm happy to add it. Just let me know.

# Documentation

I've included commits in this stack updating the only docs page that seemed relevant to me, skimming the book -- `unstable.md`.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

bearcage commented on src/cargo/core/features.rs:
```diff
@@ -371,16 +372,33 @@ impl CliUnstable {
-        let v = parts.next();
+    /// Read unstable options from a hashmap.
+    /// Intended for consuming unstable settings from config files
+    pub fn from_table(&mut self, flags: &HashMap<String, String>) -> CargoResult<()> {
```
That'd be my preference too -- I didn't add it initially for fear of making too large an unsolicited diff, but I'll see what it looks like to swap over the little bespoke deserializer table.

bearcage replied:
Hm, on closer reflection I think kicking this over to serde is orthogonal -- both the `parse` and `from_table` functions are updating-in-place, rather than deserializing to construct a new instance. I think that makes a difference here because we rely on being able to differentiate between "false because that's the default" and "explicitly set to false" to give the CLI the ability to override a config file in turning off a Zflag.

bearcage commented:
I gather it's probably fair game since we're strictly talking about unstable flags / features here, but a hard error on Zflags in config would be a break with how [mtime-on-use](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#mtime-on-use) is configured today in that stable cargo simply ignores the setting.

I can see a reasonable argument either way --

Pro: A hard error matches the behavior of setting these same flags at the command line, and reduces the chances of a nasty surprise change-in-behavior if you switch channels and recompile.

Con: A hard error means you can't leave unstable options in config and toggle back and forth between stable and nightly, or configure CI runs on nightlies to assert that a new feature does not break a build that works on stable.

Let me know how you'd like to see it work, and I'll get it changed.

alexcrichton commented:
Oh right nevermind, I forgot that we ignore it on stable/beta! That seems fine for now!

ehuss commented on src/doc/src/reference/unstable.md:
````diff
@@ -2,11 +2,22 @@
+Anything which can be configured with a Z flag can also be set in the cargo
+config file (`.cargo/config.toml`) in the `unstable` table. For example:
+
+```
````
```suggestion
```toml
```

bearcage commented:
@alexcrichton would you still like me to include a Deserialize impl in this PR, given it won't slot into the update path it's modifying?

alexcrichton commented on src/cargo/util/config/mod.rs:
```diff
@@ -742,9 +742,20 @@ impl Config {
+                // NB. It sucks to parse these twice, but doing it again here
+                //     allows the CLI to override config files for both enabling
+                //     and disabling.
+                self.unstable_flags.parse(unstable_flags)?;
```
Could this perhaps move from the beginning of the function to the end to avoid doing this twice?

bearcage replied:
It can, but it'll introduce a couple funky edge cases around bits of the config deserializer that are themselves gated on Zflags. In particular I'm thinking of [`advanced_env`](https://github.com/rust-lang/cargo/blob/master/src/cargo/util/config/de.rs#L68).

alexcrichton replied:
Ah either way is fine, if it needs to go near the beginning it's gotta go near the beginning, alas!

bearcage replied:
I gave this another go and I think (unfortunately) parsing the cli flags twice is the simplest answer for allowing explicit disables at the command line for flags that are otherwise enabled in config.

alexcrichton commented on src/cargo/util/config/mod.rs:
```diff
@@ -742,9 +742,20 @@ impl Config {
-            if let Some(val) = self.get::<Option<bool>>("unstable.mtime_on_use")? {
-                self.unstable_flags.mtime_on_use |= val;
+            if let Some(unstable_configs) =
+                self.get::<Option<HashMap<String, String>>>("unstable")?
```
A worry I have about a string=>string table is that it can be a bit surprising:

* For `build-std` I'd expect to be able to do `build-std = ['panic-abort', 'std']`
* For boolean keys I'd expect `foo = true`

Requiring a string feels a bit odd compared to what Cargo idiomatically has as configuration.

alexcrichton replied:
Another downside is that by using a `HashMap` env vars don't work. Otherwise you could do `CARGO_UNSTABLE_MTIME_ON_USE=true` for configuring Cargo.

bearcage replied:
Agreed. I much prefer getting env along for the ride as well — I'll see about making the necessary changes to move this to serde, it just might end up being a little bit more of a refactor.

alexcrichton commented on src/cargo/core/features.rs:
```diff
@@ -411,7 +430,9 @@ impl CliUnstable {
 
-        match k {
+        // Permit dashes or underscores in parsing these
+        let normalized_key = k.replace("_", "-");
```
I think it's probably best to not do this normalization to canonicalize the keys specified?

bearcage replied:
Ack — removed.

alexcrichton commented on src/cargo/core/features.rs:
```diff
@@ -371,16 +373,33 @@ impl CliUnstable {
+    /// Read unstable options from a hashmap, updating in-place.
+    /// Intended for consuming unstable settings from config files
+    pub fn update_with_table(&mut self, flags: &HashMap<String, String>) -> CargoResult<()> {
+        if !flags.is_empty() && !nightly_features_allowed() {
```
We discussed this a bit at the Cargo meeting, and I think actually the best thing to do here is to just ignore the table on the non-nightly channel. That should make it much more convenient for users testing out nightly features because all you need to do is set it in `$HOME/.cargo/config.toml` and forget about it.

bearcage replied:
Sounds good to me — I'll add tests to assert both sides of that behavior.

bearcage commented:
Just a quick update — I'm still working on this, I just hit an oddity in the Deserializer that's treating missing fields as an error even with `#[serde(default)]` on the struct. Curiously, forcing every field to use the Option deserializer first and explicitly defaulting them works, 
(a trick I shamelessly borrowed from dtolnay's answer to a totally unrelated question about serde usage elsewhere on the internet). That looks something like:

`#[serde(deserialize_with="null_default")]`
and
```rust
fn null_default<'de, D, T>(deserializer: D) -> Result<T, D::Error>
where
    D: serde::Deserializer<'de>,
    T: Deserialize<'de> + Default,
{
    let option = Option::deserialize(deserializer)?;
    Ok(option.unwrap_or_default())
}
```

I'm digging through the Deserializer now to see what's going on there.

bearcage commented:
Thinking out loud a bit here:

After talking it over with a buddy (thx @packysauce!) I think there's a bit of a layering problem in the `de` module that's preventing serde from doing defaults or deny_unknown here — when we're deserializing a struct, ConfigMapAccess is providing an iterator over the union of two sets: the fields we're interested in for this struct (the `fields` member slice), and k/v pairs available in the config table we're currently considering. As I understand them, serde MapAccess implementers are expected to serve up kv pairs of the underlying map without any particular care for what set of keys the caller is interested in. With that in mind, handling env reading inline complicates the Deserializer.

As I understand it, that means there are two problems:

1. Serde's handling for defaulting and denying-unknown will always be a noop, because we never hand back a k/v pair that the Deserialize impl didn't explicitly ask for as a field, and if we're missing a field we error out early ourselves instead of letting the derived Deserialize impl do it. This is caused by the union done in ConfigMapAccess, but fixing it would complicate point 2.
2. Environment variables don't fit super neatly in this model — if we modified ConfigMapAccess to return all k/v pairs from the config object under consideration plus all env vars which match the prefix under consideration, env vars which match _member structs_ would show up as unexpected field keys (e.g. `CARGO_UNSTABLE_SOMETABLE_SUBKEY` would show up as an unknown key in `CARGO_UNSTABLE` even if `CARGO_INSTABLE_SOMETABLE` would have been valid).

---

While I could work around this for CliUnstable by annotating every non-`Option` field with a `deserialize_with`, I think the "Right" solution here is probably a larger refactor to split the env reading out of the Deserializer impl such that Config::get<T> would try reading each source in decreasing priority order independently, with its own purpose-specific deserializer, rather than using a combined deserializer.

alexcrichton commented:
I'd definitely believe that the config bits with serde aren't exactly the best they could be, I've never really fully grok'd how to write a serde deserializer. That being said I'm happy with whatever works here, I think switching all the fields to `Option` would also be fine.

bearcage commented:
I've never totally understood the `serde` model myself, either — I'm learning my way through it on this PR. Definitely tell me if I've got it wrong!

Changes in this round:

* Rebased on master
* Added a default-to-false deserialize_with workaround to CliUnstable
* Dropped a (probably helpful) test that's broken by that change.

I'm gonna keep hacking on top of this to flip the Deserializer impl around to allow serde's `default`/`deny_unknown_fields` to do the heavy lifting, but I didn't want y'all to have to wait on me finishing that up to land this and close the linked issue, if you'd like to.

In the meantime please don't hesitate to ask if you'd like me to make some more changes to this first.

alexcrichton commented:
Looks good to me! I'd be fine landing this as-is. FWIW I don't think we want to add `deny_unknown_fields` because that'll make it painful to use this from stable Cargo when nightly adds a new feature

bearcage commented:
Just kidding, this was way simpler than I was making it out to be — the new_map ctor for ConfigMapAccess does the right thing already, it's just the struct behavior is trying to be a little too clever in how it's prefiltering fields.

I patched new_struct and dropped the workaround. If this still looks good I think it's ready to rock!

alexcrichton commented:
Oh nice! So to make sure I understand here, the previous implementation with serde was blindly assuming that all the requested fields were in the config to lookup, but this implementation is changing that to instead only iterate over fields that are actually defined in the table/env. This filtering means that we don't assume that a field requested (which could be optional) was present in the config and iterate over it (unlike before).

One thing that makes me a bit worried is the prefix-checking for the environment. We've had issues of that in the past where you've looking for `CARGO_DEBUG` but `CARGO_DEBUG_ASSERTIONS` is present in the environment. It looks like though that this won't fall down in any way that it wasn't already falling down in, right? Specifically though if we add two unstable flags `foo` and `foo-bar`, then if you set `CARGO_UNSTABLE_FOO_BAR` in the environment then will that generate an error deserializing saying that the `foo` field wasn't defined? (even though it is set to `#[serde(default)]`)

bearcage commented:
That's exactly it, yeah!

To summarize: we have 3 lists of keys:

1. The fields in the struct we're trying to deserialize
2. The list of environment variables which match our prefix
3. The CV::Table we're reading from

The previous behavior was for ConfigMapAccess to act as an iterator over `#1`, yielding the fields on the struct in order whether or not they exist in either the environment or the config file.

This patch causes ConfigMapAccess to instead iterate over the intersection of `1` and `2`, unioned with `3` — `(struct_fields ∩ env_keys) ∪ table_keys`. Effectively, it asks "What keys are present in the sources of config?" and lets the caller decide whether or not those keys are significant to them. The environment prefiltering to only matches is meant to avoid spurious "unknown field" errors on all the environment keys that would never influence the current struct anyway.

The net results of that change:

1. Missing field errors can come from derived `Deserialize` impls when a key is not present in either the environment or the config table. Previously this condition would `Err` out in the `Deserializer`, typically in one of the `deserialize_$type` functions.
2. Derived `Deserialize` impls can choose to act on extra or missing fields as they'd like. Previously, extra keys would not be reported (because we only iterated struct fields), and missing fields would be reported as `Deserializer` errors. This means the `#[serde()]` annotations for `default` and `deny_unknown_fields` should work as expected now, with one exception (next point).
3. Environment variables which match a prefix will be reported. As you've pointed out, this does mean that env vars for sub-objects in config will show up in the set of keys for the containing object. In the case where either `#[serde(default)]` is set or no annotation is provided, this should be identical to the current behavior because the derived `Deserialize` will wind up simply skipping those env entries as unknown fields. The new failure mode is in combining env sub-object keys  with `deny_unknown_fields` — I think that will break in a potentially surprising way in that it'll cause an error for an otherwise-valid subkey setting.

I'll add some tests around that behavior now to make sure I've got the combination of behaviors right here — it occurs to me that we might even be able to fix that by doing exact matches on env keys in ConfigMapAccess, since subkeys shouldn't _need_ to be reported there.

bearcage commented:
Looks like we've already got tests covering most of those cases!

* `config::struct_with_opt_inner_struct` checks that setting an env key for a nested struct results in trying to deserialize the nested struct rather than defaulting it. This test correctly fails if I switch ConfigMapAccess to `eq` instead of `starts_with` on env keys, because that'd lead to skipping a member struct if there's no other indication that data is present to deserialize it.
* `config::overlapping_env_config` asserts that optional keys with overlapping prefixes are correctly independent.

I've added a couple more using default annotations based on the existing ones. These demonstrate the failure mode you mentioned pretty clearly. I think that might be a somewhat unavoidable limitation of the mapping between environment variables and nested configs, unfortunately, but you're correct that this patch fails in the same way as the existing code in the overlapping_env scenario (and doesn't fail in any new ways).

bearcage commented:
On second thought — it seems preferable to error out when we encounter an ambiguous situation like this, versus incorrectly defaulting an inner field. I'll revert the env prefix tweak and patch the new tests accordingly.

bearcage commented:
Ok, with that done — your summary of the change is entirely correct, and now we've got tests asserting that.

alexcrichton commented:
Ok this all seems resonable to me, thanks again for helping to explain all this! It sounds like we'll want to be careful to avoid overlapping prefixes where possible until we fix these issues (if they're fixable at all, really), but that's something we can try to remain vigilant about in the future. 

Thanks again for your work here!

@bors: r+

bors commented:
:pushpin: Commit d035daae4e26bfe144ceff41f4c9a0bb11fea605 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d035daae4e26bfe144ceff41f4c9a0bb11fea605 -->
<!-- homu: {"type":"Approved","sha":"d035daae4e26bfe144ceff41f4c9a0bb11fea605","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit d035daae4e26bfe144ceff41f4c9a0bb11fea605 with merge a02e7655e7638a14e02584117a8b489e00ed8b0f...
<!-- homu: {"type":"BuildStarted","head_sha":"d035daae4e26bfe144ceff41f4c9a0bb11fea605","merge_sha":"a02e7655e7638a14e02584117a8b489e00ed8b0f"} -->

bearcage commented:
Happy to help - thanks for the reviews!

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=34196)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=34196","builder_name":"checks-azure"} -->

bearcage commented:
Test failure, bummer.

I couldn't find a way to pull the proptest codegen'd case off the builder host, so I've left my workstation running a copy of the failing proptest (`prop_limited_independence_of_irrelevant_alternatives`) per core until I get a reproducer. Fingers crossed it doesn't take toooooo long.

ehuss commented:
@bors retry
https://github.com/rust-lang/cargo/issues/6258#issuecomment-656404282

bors commented:
:hourglass: Testing commit d035daae4e26bfe144ceff41f4c9a0bb11fea605 with merge f35ebafc74fed12ba7c5f7c497a7e751889c11dd...
<!-- homu: {"type":"BuildStarted","head_sha":"d035daae4e26bfe144ceff41f4c9a0bb11fea605","merge_sha":"f35ebafc74fed12ba7c5f7c497a7e751889c11dd"} -->

bearcage commented:
thx ehuss!

I just got a repro case locally, but skimming the linked thread it looks like y'all already have a handle on where the edge case proptest is turning up is so I'll let it lie.

bors commented:
:sunny: Test successful - [checks-azure](https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=34205)
Approved by: alexcrichton
Pushing f35ebafc74fed12ba7c5f7c497a7e751889c11dd to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/473adcf1-6656-4c7e-9337-c393e7e0431c/_build/results?buildId=34205"},"merge_sha":"f35ebafc74fed12ba7c5f7c497a7e751889c11dd"} -->

Eh2406 commented:
Technically there have to be an infinite supply of edge cases. The problem is NP-Complete so there have to be exponencial cases. So far (at least recently) the proptest are finding particularly un-realistic examples and my investigations have not found ways to make them fast that are worth the additional code complexity. But each one is a new adventure, that can be completely unrelated.

phil-opp commented:
I just tried this feature with the latest nightly and it seems to work great overall. Thanks a lot @bearcage for implementing this!

One minor thing I noticed was that the `-Zbuild=core` command line argument behaves slightly different than setting `unstable.build-std = ["core"]` in a `.cargo/config` file. The former seems to automatically derive that `compiler_builtins` is also needed when building `core`, while the latter seems not. With the config option, a "can't find crate for `compiler_builtins`" error occurs unless I add `compiler_builtins` to the `unstable.build-std` list. This isn't a problem for me, but I wanted to report it in case this is unintended.

alexcrichton commented:
@phil-opp looks like [that's a bug](https://github.com/rust-lang/cargo/pull/8491)

phil-opp commented:
Thanks a lot for the quick investigation and fix!

simnalamburt commented:
Thanks to this PR, now we can use the rust-analyzer with the unstable cargo option! Yay
